### PR TITLE
Add missing literal types from previous version

### LINKED ISSUE #77: SIGSEGV: so-called oddball literals
With:

- libcon4m b4fdc8e2e70b0ef18a554562702e2db1e0c9aff0
- x86_64
- Linux 6.9.7
- clang 17.0.6

`c4test` no longer crashes (see https://github.com/crashappsec/libcon4m/issues/54) with these:

```text
x = 1'foo
```

```text
x = "1"'foo
```
 
but it does crash for cases like these:

```text
x = "1"'sz
```

```text
x = "2024-01-01"'date
```

```text
x = "https://example.com"'url
```

with traces like:

```text
Invalid read of size 8
  at 0x1480AC: c4m_parse_simple_lit (../src/con4m/literals.c:274)
  by 0x16162B: simple_lit (../src/con4m/compiler/parse.c:890)
  by 0x1645DE: literal (../src/con4m/compiler/parse.c:2638)
  by 0x1660F3: expression_start (../src/con4m/compiler/parse.c:3239)
  by 0x160BF4: expression (../src/con4m/compiler/parse.c:3794)
  by 0x160C85: assign (../src/con4m/compiler/parse.c:3060)
  by 0x15E3F8: module (../src/con4m/compiler/parse.c:4222)
  by 0x15D93B: c4m_parse (../src/con4m/compiler/parse.c:4359)
  by 0x158E77: c4m_initial_load_one (../src/con4m/compiler/compile.c:624)
  by 0x158D0E: c4m_perform_module_loads (../src/con4m/compiler/compile.c:741)
  by 0x159291: c4m_compile_from_entry_point (../src/con4m/compiler/compile.c:1052)
  by 0x116D5A: test_compiler (../src/tests/test.c:427)
Address 0x48 is not stack'd, malloc'd or (recently) free'd

Process terminating with default action of signal 11 (SIGSEGV): dumping core
```

This is probably just not yet implemented, but let's track it here.

The literals are registered here:

https://github.com/crashappsec/libcon4m/blob/b4fdc8e2e70b0ef18a554562702e2db1e0c9aff0/src/con4m/literals.c#L147-L220

### PR DESCRIPTION
**Pull request by viega**
Re-implemented the 'other' literal types:
Datetime
Time
Date
Size
Duration

URLs are done not as a separate type, but as a literal modifier that calls CURL's URL validation (loose as it is).

Added an option to turn on / off static binding of calls exposed via the FFI. If not set, it's on if glibc is off on linux, and off by default on mac.

Also, for simple assignment ONLY, if the assignment is to a variable with an explicitly declared type (and that type is a simple literal), then we will infer the litmod type, picking the first one added associated with the type.

Closes #77

### CODE REVIEW

**Comment by viega** [comment]
Certainly don't review until I fix the Linux tests; was surprised it didn't work there, but I'll take a look first.

**Comment by viega** [comment]
@ee7 Fixed, feel free.

**Review by ee7 — COMMENTED** [review]
Tests pass locally, and no new warnings or ASan/UBSan output.

Should this PR close ticket https://github.com/crashappsec/libcon4m/issues/77? The examples in that ticket no longer crash.

There's some strange time formatting at the end of the tests, e.g. from the [CI log](https://github.com/crashappsec/libcon4m/actions/runs/10081511491/job/27873694365#step:11:1430):

```
Now (Time since the Unix Epoch)
54 years, 31 weeks, 1 day, 17 hours, 47 minutes, 48 seconds 740647 msec, 390 nsec
Monotonic Timestamp (POSIX; no guaranteed start time)
14 minutes, 39 seconds 448521 msec, 937 nsec
Process CPU
34525 msec, 521 nsec
Thread CPU
34702 msec, 270 nsec
System Uptime (generally excludes sleep)
14 minutes, 39 seconds 449223 msec, 474 nsec
Program wall clock (all proceeding tests till now)
2 seconds 46167 msec, 936 nsec
```

5 or 6 digits of `msec` looks wrong. Should this:

- If wanting to specify seconds, microseconds, and nanoseconds: change `msec` to `μs` or `us`?
- If wanting to specify seconds, milliseconds, microseconds, and nanoseconds: add the missing one?

File: tests/size.c4m [hunk 1]
```diff
@@ -0,0 +1,21 @@
+"""
+Size matters.
+"""
+"""
+$output:
+10 MiB 3 Bytes
+10 Mb 1 Bytes
+1 Kb 23 Bytes
+"""
+s1 = "10485763 b"'sz
```
[the remark this answers is not in the document]
Comment (viega) [reply tests/size.c4m:10]: No, and beyond what I just pushed, I am not addressing anything else here any time soon, it is far from being a priority.

**Review by ee7 — APPROVED** [review]
Should this PR close ticket https://github.com/crashappsec/libcon4m/issues/77? The examples in that ticket no longer crash.

**Pull request merged**